File: ios/Classes/Extension/PPLefuBleExtensionData.swift
//
//  PPLefuBleDataExtension.swift
//  Pods
//
//  Created by lefu on 2025/4/16
//  


import Foundation
import PPBaseKit
import PPBluetoothKit
import Flutter

extension PPLefuBleConnectManager {
    
    func convert180A(model:PPBluetooth180ADeviceModel)->[String:Any] {
        let dict:[String:Any?] = [
            "modelNumber":model.modelNumber,
            "firmwareRevision":model.firmwareRevision,
            "softwareRevision":model.softwareRevision,
            "hardwareRevision":model.hardwareRevision,
            "serialNumber":model.serialNumber,
            "manufacturerName":model.manufacturerName
        ]
        
        
        let filtedDict = dict.compactMapValues { $0 }
        
        return filtedDict
    }
    
    func convertDeviceDict(_ device:PPBluetoothAdvDeviceModel)->[String:Any] {
        let dict:[String:Any] = [
            "deviceSettingId":device.deviceSettingId,
            "deviceMac":device.deviceMac,
            "deviceName":device.deviceName,
            "customDeviceName":device.customDeviceName,
            "devicePower":device.devicePower,
            "rssi":device.rssi,
            "deviceType":device.deviceType.rawValue,
            "deviceProtocolType":device.deviceProtocolType.rawValue,
            "deviceCalculateType":device.deviceCalcuteType.rawValue,
            "deviceAccuracyType":device.deviceAccuracyType.rawValue,
            "devicePowerType":device.devicePowerType.rawValue,
            "deviceConnectType":device.deviceConnectType.rawValue,
            "deviceFuncType":device.deviceFuncType.rawValue,
            "deviceUnitType":device.deviceUnitList,
            "peripheralType":device.peripheralType.rawValue,
            "sign":device.sign,
            "advLength":device.advLength,
            "macAddressStart":device.macAddressStart,
            "standardType":device.standardType.rawValue
        ]
        
        return dict
    }
    
    func convertMeasurementDict(_ model:PPBluetoothScaleBaseModel)->[String:Any] {
        
        let dateTimeInterval = Int(model.dateTimeInterval * 1000)
        let memberId = model.memberId
        let isPowerOff = model.isPowerOff
        
        let dict:[String:Any?] = [
            "weight":model.weight,
            "impedance":model.impedance,
            "impedance100EnCode":model.impedance100EnCode,
            "isHeartRating":model.isHeartRating,
            "heartRate":model.heartRate,
            "isOverload":model.isOverload,
            "isPlus":model.isPlus,
            "measureTime":dateTimeInterval,
            "memberId":memberId,
            "footLen":model.footLen,
            "unit":model.unit.rawValue,
            "z100KhzLeftArmEnCode":model.z100KhzLeftArmEnCode,
            "z100KhzLeftLegEnCode":model.z100KhzLeftLegEnCode,
            "z100KhzRightArmEnCode":model.z100KhzRightArmEnCode,
            "z100KhzRightLegEnCode":model.z100KhzRightLegEnCode,
            "z100KhzTrunkEnCode":model.z100KhzTrunkEnCode,
            "z20KhzLeftArmEnCode":model.z20KhzLeftArmEnCode,
            "z20KhzLeftLegEnCode":model.z20KhzLeftLegEnCode,
            "z20KhzRightArmEnCode":model.z20KhzRightArmEnCode,
            "z20KhzRightLegEnCode":model.z20KhzRightLegEnCode,
            "z20KhzTrunkEnCode":model.z20KhzTrunkEnCode,
            "isPowerOff":isPowerOff
                
        ]
        
        let filtedDict = dict.compactMapValues { $0 }
        
        return filtedDict
    }
    
    func sendMeasureData(_ model:PPBluetoothScaleBaseModel, advModel: PPBluetoothAdvDeviceModel, measureState:Int) {
        
        let deviceDict:[String:Any] = self.convertDeviceDict(advModel)
        let dataDict:[String:Any] = self.convertMeasurementDict(model)
        
        self.loggerStreamHandler?.event?("测量状态:\(measureState)")
        
        let dict:[String:Any] = [
            "measurementState":measureState,
            "device":deviceDict,
            "data":dataDict
        ]
        
        self.measureStreamHandler?.event?(dict)
    }
    
    func sendHistoryData(_ models:[PPBluetoothScaleBaseModel]) {
        
        var array:[[String:Any]] = []
        
        for model in models {
            let dict = self.convertMeasurementDict(model)
            array.append(dict)
        }

        let dataList = array
        
        
        self.loggerStreamHandler?.event?("历史数据-数量:\(dataList.count)")
        let dict = [
            "dataList":dataList
        ]
        self.historyStreamHandler?.event?(dict)
    }
    
    func sendBlePermissionState(state:PPBluetoothState) {
        
        var stateValue:Int = 0
        if state == .unauthorized {
            stateValue = 1
        } else if state == .poweredOn {
            stateValue = 2
        } else if state == .poweredOff {
            stateValue = 3
        }
        
        let dict:[String:Any] = [
            "state":stateValue
        ]
        
        self.blePermissionStreamHandler?.event?(dict)
    }
    
    func sendWIFIResult(isSuccess:Bool, sn:String?, errorCode:Int?, callBack: @escaping FlutterResult) {
        
        let dict:[String:Any?] = [
            "success":isSuccess,
            "errorCode":errorCode,
            "sn":sn
        ]
        
        let filtedDict = dict.compactMapValues { $0 }
        
        callBack(filtedDict)
    }
    
    func sendWIFISSID(_ ssId:String?, isConnectWIFI:Bool, callBack: @escaping FlutterResult) {
        
        let dict:[String:Any?] = ["ssId": ssId, "isConnectWIFI":isConnectWIFI]
        let filtedDict = dict.compactMapValues { $0 }
        
        callBack(filtedDict)
    }
    
    func sendWifiList(_ wifiList:[PPWifiInfoModel], callBack: @escaping FlutterResult) {
        
        var array = [String]()
        for model in wifiList {
            array.append(model.ssid)
        }
        let dict:[String:Any] = ["wifiList":array]
        callBack(dict)
    }
    
    func sendWifiOTA(isSuccess:Bool, errorCode:Int, callBack: @escaping FlutterResult) {

        let dict:[String:Any] = [
            "isSuccess":isSuccess,
            "errorCode":errorCode
        ]
        
        callBack(dict)
    }
    
    func sendCommonState(_ state:Bool, callBack: @escaping FlutterResult) {
        
        let dict:[String:Any] = [
            "state":state
        ]
        
        callBack(dict)
    }
    
    func sendDfuResult(progress:Float, isSuccess:Bool) {
        
        let dict:[String:Any] = [
            "progress":progress,
            "isSuccess":isSuccess,
        ]
        
        self.dfuStreamHandler?.event?(dict)
    }
    
    func sendScanState(scaning:Bool) {
        let code:Int = scaning ? 1 : 0
        self.scanStateStreamHandler?.event?(["state":code])
    }
    
    
    func sendKitchenData(_ model:PPBluetoothScaleBaseModel, advModel: PPBluetoothAdvDeviceModel, measureState:Int) {
        
        let deviceDict:[String:Any] = self.convertDeviceDict(advModel)
        let dataDict:[String:Any] = self.convertMeasurementDict(model)
        
        self.loggerStreamHandler?.event?("厨房秤-测量状态:\(measureState)")
        
        let dict:[String:Any] = [
            "measurementState":measureState,
            "device":deviceDict,
            "data":dataDict
        ]
        
        self.kitchenStreamHandler?.event?(dict)
    }
    
}

File: android/src/main/kotlin/com/example/pp_bluetooth_kit_flutter/extension/PPLefuBleExtensionData.kt
package com.example.pp_bluetooth_kit_flutter.extension

import com.example.pp_bluetooth_kit_flutter.PPLefuBleConnectManager
import com.example.pp_bluetooth_kit_flutter.util.PPBluetoothState
import com.lefu.ppbase.PPBodyBaseModel
import com.lefu.ppbase.PPDeviceModel
import com.lefu.ppbase.vo.PPScaleStatePowerType
import com.peng.ppscale.PPBluetoothKit
import com.peng.ppscale.util.DateUtil
import com.peng.ppscale.vo.LFFoodScaleGeneral
import io.flutter.plugin.common.MethodChannel.Result


/**
 * PPLefuBleConnectManager的数据处理扩展
 * Created by lefu on 2023/4/16
 */
fun PPLefuBleConnectManager.convert180A(model: PPDeviceModel?): Map<String, Any> {
    val map = mutableMapOf<String, Any?>()
    map["modelNumber"] = model?.modelNumber
    map["firmwareRevision"] = model?.firmwareVersion
    map["softwareRevision"] = model?.softwareVersion
    map["hardwareRevision"] = model?.hardwareVersion
    map["serialNumber"] = model?.serialNumber
    map["manufacturerName"] = model?.manufacturerName

    return map.filterValues { it != null } as Map<String, Any>
}

fun PPLefuBleConnectManager.convertDeviceDict(device: PPDeviceModel): Map<String, Any> {
    val map = mutableMapOf<String, Any?>()
    map["deviceSettingId"] = device.deviceSettingId
    map["deviceMac"] = device.deviceMac
    map["deviceName"] = device.deviceName
    map["devicePower"] = device.devicePower
    map["rssi"] = device.rssi
    map["deviceType"] = device.deviceType.ordinal
    map["deviceProtocolType"] = device.deviceProtocolType.getType()
    map["deviceCalculateType"] = device.deviceCalcuteType.getType()
    map["deviceAccuracyType"] = device.deviceAccuracyType.getType()
    map["devicePowerType"] = device.devicePowerType.getType()
    map["deviceConnectType"] = device.deviceConnectType.getType()
    map["deviceFuncType"] = device.deviceFuncType
    map["deviceUnitType"] = device.deviceUnitType
    map["peripheralType"] = device.getDevicePeripheralType().ordinal
    map["sign"] = device.sign
    map["advLength"] = device.advLength
    map["macAddressStart"] = device.macAddressStart
    map["standardType"] = device.standardType
    map["productModel"] = device.productModel
    val deviceConfig = PPBluetoothKit.getDeviceConfigVo(device.deviceSettingId);
    map["brandId"] = (deviceConfig?.brandId ?: "0").toInt()
    map["imgUrl"] = device.imgUrl
    map["avatarType"] = deviceConfig?.avatarType ?: 0
    map["customDeviceName"] = deviceConfig?.customDeviceName ?: ""

    return map.filterValues { it != null } as Map<String, Any>
}

fun PPLefuBleConnectManager.convertMeasurementDict(model: PPBodyBaseModel): Map<String, Any> {
    val map = mutableMapOf<String, Any?>()
    if (model.dateStr.isNullOrEmpty()) {
        val dateTimeInterval = System.currentTimeMillis()
        map["measureTime"] = dateTimeInterval
    } else {
        map["measureTime"] = DateUtil.stringToLong(model.dateStr)
    }
    val memberId = model.memberId

    map["weight"] = model.weight
    map["impedance"] = model.impedance
    map["impedance100EnCode"] = model.ppImpedance100EnCode
    map["isHeartRating"] = model.isHeartRating
    map["heartRate"] = model.heartRate
    map["isOverload"] = model.isOverload
    map["memberId"] = memberId
    map["footLen"] = model.footLen
    map["unit"] = model.unit?.type
    map["z100KhzLeftArmEnCode"] = model.z100KhzLeftArmEnCode
    map["z100KhzLeftLegEnCode"] = model.z100KhzLeftLegEnCode
    map["z100KhzRightArmEnCode"] = model.z100KhzRightArmEnCode
    map["z100KhzRightLegEnCode"] = model.z100KhzRightLegEnCode
    map["z100KhzTrunkEnCode"] = model.z100KhzTrunkEnCode
    map["z20KhzLeftArmEnCode"] = model.z20KhzLeftArmEnCode
    map["z20KhzLeftLegEnCode"] = model.z20KhzLeftLegEnCode
    map["z20KhzRightArmEnCode"] = model.z20KhzRightArmEnCode
    map["z20KhzRightLegEnCode"] = model.z20KhzRightLegEnCode
    map["z20KhzTrunkEnCode"] = model.z20KhzTrunkEnCode

    if (model.scaleState?.powerType != PPScaleStatePowerType.PP_SCALE_STATE_POWER_OFF) {
        map["isPowerOff"] = true
    } else {
        map["isPowerOff"] = true
    }
    return map.filterValues { it != null } as Map<String, Any>
}

fun PPLefuBleConnectManager.convertMeasurementDictFood(model: LFFoodScaleGeneral): Map<String, Any> {

    val map = mutableMapOf<String, Any?>()
    map["weight"] = model.lfWeightKg
    map["isPlus"] = model.thanZero
    map["measureTime"] = System.currentTimeMillis()
    map["unit"] = model.unit

    return map.filterValues { it != null } as Map<String, Any>
}


fun PPLefuBleConnectManager.sendMeasureData(model: PPBodyBaseModel, advModel: PPDeviceModel, measureState: Int) {
    val deviceDict = this.convertDeviceDict(advModel)
    val dataDict = this.convertMeasurementDict(model)

    this.loggerStreamHandler?.sendEvent("测量状态:$measureState")

    val dict = mapOf(
        "measurementState" to measureState,
        "device" to deviceDict,
        "data" to dataDict
    )

    this.measureStreamHandler?.sendEvent(dict)
}

fun PPLefuBleConnectManager.sendHistoryData(models: List<PPBodyBaseModel>) {
    val array = mutableListOf<Map<String, Any>>()

    for (model in models) {
        val dict = this.convertMeasurementDict(model)
        array.add(dict)
    }

    val dataList = array

    this.loggerStreamHandler?.sendEvent("历史数据-数量:${dataList.size}")
    val dict = mapOf(
        "dataList" to dataList
    )
    this.historyStreamHandler?.sendEvent(dict)
}

fun PPLefuBleConnectManager.sendBlePermissionState(state: PPBluetoothState) {
    var stateValue = 0
    if (state == PPBluetoothState.UNAUTHORIZED) {
        stateValue = 1
    } else if (state == PPBluetoothState.POWERED_ON) {
        stateValue = 2
    } else if (state == PPBluetoothState.POWERED_OFF) {
        stateValue = 3
    }

    val dict = mapOf(
        "state" to stateValue
    )

    this.blePermissionStreamHandler?.sendEvent(dict)
}

fun PPLefuBleConnectManager.sendWIFIResult(isSuccess: Boolean, sn: String?, errorCode: Int?, callBack: Result) {
    val map = mutableMapOf<String, Any?>()
    map["success"] = isSuccess
    map["errorCode"] = errorCode
    map["sn"] = sn

    val filteredMap = map.filterValues { it != null }

    callBack.success(filteredMap)
}

fun PPLefuBleConnectManager.sendWIFISSID(ssId: String?, isConnectWIFI: Boolean, callBack: Result) {
    val map = mutableMapOf<String, Any?>()
    map["ssId"] = ssId
    map["isConnectWIFI"] = isConnectWIFI

    val filteredMap = map.filterValues { it != null }

    callBack.success(filteredMap)
}

/**
 * 0-成功;
 * 1-设备已在升级中不能再次启动升级;
 * 2-设备低电量无法
 * 启动升级;
 * 3-未配网
 * 4-充电中
 */
fun PPLefuBleConnectManager.sendWifiOTA(isSuccess: Boolean, errorCode: Int, callBack: Result) {
    val dict = mapOf(
        "isSuccess" to isSuccess,
        "errorCode" to errorCode
    )

    callBack.success(dict)
}

fun PPLefuBleConnectManager.sendCommonState(state: Boolean, callBack: Result?) {
    val dict = mapOf(
        "state" to state
    )

    callBack?.success(dict)
}

fun PPLefuBleConnectManager.sendDfuResult(progress: Float, isSuccess: Boolean) {
    val dict = mapOf(
        "progress" to progress,
        "isSuccess" to isSuccess
    )

    this.dfuStreamHandler?.sendEvent(dict)
}

fun PPLefuBleConnectManager.sendScanState(scaning: Boolean) {
    val code = if (scaning) 1 else 0
    this.scanStateStreamHandler?.sendEvent(mapOf("state" to code))
}

fun PPLefuBleConnectManager.sendKitchenData(model: LFFoodScaleGeneral, deviceModel: PPDeviceModel, measureState: Int) {
    val deviceDict = this.convertDeviceDict(deviceModel)
    val dataDict = this.convertMeasurementDictFood(model)

    this.loggerStreamHandler?.sendEvent("厨房秤-测量状态:$measureState")

    val dict = mapOf(
        "measurementState" to measureState,
        "device" to deviceDict,
        "data" to dataDict
    )

    this.kitchenStreamHandler?.sendEvent(dict)
}